rfcs/006-package-management.md

```diff
@@ -0,0 +1,243 @@
+on the advantages and disadvantages of inline imports.
+
+### Alternative: toml manifest
+
+Maybe the manifest should be in some plain-data format like toml. This would
+be easier to modify programmatically.
+
+### Alternative: shorthand for registry imports
+
+Cargo allows a shorthand like
+
+```toml
+    "github/tweag/foo" = "1.2.0"
+```
+
+instead of
+
+```nickel
+    foo = 'Index { package = "github/tweag/foo", version = "1.2.0" }
+```
+
+Since we expect registry imports to be the common case, maybe it's worth having
+a shorthand?
+
+### Question: manifest file name
+
+The name `package.ncl` was chosen to be similar to npm's `package.json` or
+stack's `package.yaml`. One problem with this is that it could be confused for
+a nickel source file. Another possibility would be to use an extension-based
+name like `<package-name>.cabal`. Or just a stranger name that's less likely to
+conflict with something real.
+
+## Import statements
+
+The manifest file assigns a name to each dependency; to import the dependency
+named `foo` you simply write `import foo`. That is, an `import` statement either
+takes a string in quotes -- in which case it imports a path -- or an identifier
+without quote -- in which case it imports a package.
+
+The `import foo` expression evaluates to the contents of `main.ncl` in `foo`'s
+root directory.
+
+### Question: other entry points?
+
+We've hardcoded `main.ncl` as the entry point of every package, but what if
+they want to expose multiple entry points? For example, node allows a package's
+manifest to specify the entry point(s). This is probably not very important to
+support, as you can just put
+
+```nickel
+{
+  other = import "./other.ncl",
+  blah = import "./blah.ncl",
+}
+```
+
+in your package's `main.ncl`, to provide "other" and "blah" as other entry points.
+
+## Kinds of dependencies
+
+Where can dependencies come from? Dhall allows imports from arbitrary urls. Nix
+supports fetching from a variety of VCSs, paths, and archive formats.
+
+We'll support dependencies from
+
+- a central registry, that can identify packages by name and version number.
+  This should be the most common method of importing packages, like `crates.io`
+  in rust.
+- git repositories (either from HEAD, or from branches, tags, and revisions
+  specified by hashes). This allows for easy use of unpublished packages,
+  including in-development versions.
+- paths (relative or absolute). This allows for easy use of different packages
+  within the same repository, or for temporary patching of published packages.
+
+We require packages to have their own manifest file (at the package root), even
+if they don't import dependencies.
+
+## Lock files
+
+In order to ensure reproducibility across time and across machines, we build
+a lock-file (if there isn't yet one) when running `nickel eval` or `nickel
+export`. The lock-file specifies the exact versions of all (transitive)
```

Comment:
If I understand correctly, this means that we entirely tie the interpreter and the package manager (of course, the interpreter needs to have some knowledge of the package manager, because it needs to find the file on the disk - but that interface can be very simple and light weight, as in I give you a name, you give me a concrete path). I wonder if we should also consider separating them, with Nickel complaining about a missing lockfile (and asking to run `metalist` or whatnot). Although it does have a cost in term of user experience

Reply:
So the current code structure of the prototype is:
- `nickel-lang-core` gets a new `PackageMap` type, which basically just answers the question "if package `x` imports package `y`, what path should it come from"
- there's a new crate `nickel-lang-package` that has all the logic of package fetching and version resolution. It depends on `nickel-lang-core`, and is capable of producing a `PackageMap` from the resolution.
- `nickel-lang-cli` adds a dep on `nickel-lang-package` to integrate the package management part. I've put some package-management tools under `nickel package xyz`, and added package support to `eval` and `export`

As an alternative, `nickel-lang-cli` could remain untouched and we could put the package-management utilities under, say, `metalist`.

There is one little annoyance related to path dependencies, because the lock-file doesn't currently include their recursive dependencies. This means that we (currently) need to do some resolution on-the-fly just before evaluation. So with the current state of things it would be hard to support the interface

```console
metalist lock
nickel export foo.ncl
```

and we'd probably need an interface that wraps nickel, like `metalist export foo.ncl`

Reply:
It was intentional, rather than accidental, but it's possible I'm falling into the trap of copying cargo's behavior too much...

There are two inherent problems with path dependencies: they can change at any time, and they might not have the same contents across machines. This means that "locking" them doesn't do much to help reproducibility. It also means that the path dependencies are just as mutable as the top-level manifest -- every time we read and evaluate the top-level manifest we also need to read and evaluate all the path-dependency manifests. So I guess there are two possibilities:

- with package management integrated into the cli (as in the current prototype) path dependencies should not be locked, and their manifests should be re-read and re-resolved on every `nickel eval`
- with package management as a separate tool, path dependencies should be locked (because `nickel eval` will consume the lock-file and needs to know about them). Every time `metallist` is run, the path dependencies should be re-resolved. If the path dependencies are modified in between running `metallist` and running `nickel`, you'll just get an error. (Same will happen if you modify the top-level manifest.)

Reply:
We could ask for confirmation before changing the resolution (or print out the changes)